core/trino-main/src/main/java/io/trino/operator/window/PatternRecognitionPartition.java

```diff
@@ -0,0 +1,531 @@
+            measureComputation.getType().appendTo(result, 0, pageBuilder.getBlockBuilder(channel));
+            channel++;
+        }
+        // window functions have frame consisting of all rows of the match
+        for (FramedWindowFunction framedFunction : windowFunctions) {
+            framedFunction.getFunction().processRow(
+                    pageBuilder.getBlockBuilder(channel),
+                    peerGroupStart - partitionStart,
+                    peerGroupEnd - partitionStart - 1,
+                    patternStart - partitionStart,
+                    patternStart + labels.length() - 1 - partitionStart);
+            channel++;
+        }
+    }
+
+    private void outputAllRowsPerMatch(PageBuilder pageBuilder, MatchResult matchResult)
+    {
+        // window functions are not allowed with ALL ROWS PER MATCH
+        checkState(windowFunctions.isEmpty(), "invalid node: window functions specified with ALL ROWS PER MATCH");
+
+        ImmutableArray labels = matchResult.getLabels();
+
+        // determine positions to output
+        List<Integer> positionsToOutput;
+        ImmutableArray exclusions = matchResult.getExclusions();
+
+        if (exclusions.length() == 0) {
+            positionsToOutput = range(currentPosition, currentPosition + labels.length())
+                    .boxed()
+                    .collect(toImmutableList());
+        }
+        else {
+            IntStream relativePositions = range(0, exclusions.get(0));
+            int index = 1;
+            while (index < exclusions.length() - 2) {
+                relativePositions = concat(relativePositions, range(exclusions.get(index), exclusions.get(index + 1)));
+                index += 2;
+            }
+            relativePositions = concat(relativePositions, range(exclusions.get(index), labels.length()));
+            positionsToOutput = relativePositions
+                    .map(relativePosition -> currentPosition + relativePosition)
+                    .boxed()
+                    .collect(toImmutableList());
+        }
+
+        for (int i : positionsToOutput) {
+            // copy output channels
+            pageBuilder.declarePosition();
+            int channel = 0;
+            while (channel < outputChannels.length) {
+                pagesIndex.appendTo(outputChannels[channel], i, pageBuilder.getBlockBuilder(channel));
+                channel++;
+            }
+            // compute measures from the current position (the position from which measures are computed matters in RUNNING semantics)
+            for (MeasureComputation measureComputation : measures) {
+                Block result = measureComputation.compute(i, labels, partitionStart, partitionEnd, currentPosition, matchNumber, windowIndex);
+                measureComputation.getType().appendTo(result, 0, pageBuilder.getBlockBuilder(channel));
+                channel++;
+            }
+        }
```

Comment:
This method can be simplified to:

```java
        int start = 0;
        for (int index = 0; index < exclusions.length(); index += 2) {
            int end = exclusions.get(index);

            for (int i = start; i < end; i++) {
                outputRow(pageBuilder, labels, currentPosition + i);
            }

            start = exclusions.get(index + 1);
        }

        for (int i = start; i < labels.length(); i++) {
            outputRow(pageBuilder, labels, currentPosition + i);
        }
```

You'll need to extract the body of the for loop into an `outputRow` function.